scanner/src/main/kotlin/ArtifactoryCache.kt

```diff
@@ -52,44 +55,85 @@ class ArtifactoryCache(
-                            "Retrieved $cachePath from local cache."
-                        } else {
-                            "Downloaded $cachePath from Artifactory cache."
-                        }
+                    if (response.cacheResponse() != null) {
+                        log.info { "Retrieved $cachePath from local cache." }
                     } else {
+                        log.info { "Downloaded $cachePath from Artifactory cache." }
+                    }
+
+                    return yamlMapper.readValue(tempFile, ScanResults::class.java)
+                } else {
+                    log.info {
                         "Could not get $cachePath from Artifactory cache: ${response.code()} - " +
                                 response.message()
                     }
-
-                    log.info { message }
                 }
             }
         } catch (e: IOException) {
             e.showStackTrace()
 
             log.warn { "Could not get $cachePath from Artifactory cache: ${e.message}" }
+        }
 
-            return false
+        return ScanResults(id, emptyList())
+    }
+
+    override fun read(id: Identifier, scannerSpecification: ScannerSpecification): ScanResults {
+        val scanResults = read(id)
+        val filteredResults = scanResults.results.filter {
+            it.scanner == scannerSpecification
+        }
+
+        if (filteredResults.isEmpty() && scanResults.results.isNotEmpty()) {
+            log.info { "No cached scan results found for '$scannerSpecification', ignoring cached scan results for " +
+                    "scanners: ${scanResults.results.map { it.scanner }}" }
         }
+
+        return ScanResults(id, filteredResults)
     }
 
-    override fun write(pkg: Package, source: File): Boolean {
-        val cachePath = cachePath(pkg, source)
+    override fun add(id: Identifier, scanResult: ScanResult): Boolean {
+        if (scanResult.summary.fileCount == 0) {
+            // Do not cache empty scan results. It is likely that something went wrong when they were created, and if
+            // not, it is cheap to re-create them.
+
+            log.info { "Not caching scan result for '$id' because no files were scanned." }
+
+            return false
+        }
+
+        if (scanResult.provenance.sourceArtifact == null && scanResult.provenance.vcsInfo == null) {
+            // Do not cache scan results without provenance information, because they cannot be assigned to the revision
+            // of the package source code later.
 
-        log.info { "Writing scan results to Artifactory cache: $cachePath" }
+            log.info { "Not caching scan result for '$id' because no provenance information is available." }
+
+            return false
+        }
+
+        val scanResults = read(id).run {
+            ScanResults(id, results.toMutableList().apply { add(scanResult) })
+        }
+
+        val tempFile = createTempFile("scan-results-")
```

Comment:
Maybe:

        val tempFile = createTempFile("scan-results-").also {
            yamlMapper.writeValue(it, scanResults)
        }

Reply:
I prefer to keep it as is, because `tempFile` is used also below.